flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java

```diff
@@ -301,81 +306,388 @@ public void testProducerFailedException() throws Exception {
+			// Release all the buffer resources once exception
+			if (!inputChannel.isReleased()) {
+				inputChannel.releaseAllResources();
+			}
+
+			networkBufferPool.destroyAllBufferPools();
+			networkBufferPool.destroy();
+
+			executor.shutdown();
+		}
 	}
 
 	/**
-	 * Tests {@link RemoteInputChannel#releaseAllResources()}, verifying the exclusive segments are
-	 * recycled to global pool via input gate and no resource leak.
+	 * Tests to verify that there is no race condition with two things running in parallel:
+	 * requesting floating buffers on sender backlog and some other thread recycling
+	 * floating or exclusive buffers.
 	 */
 	@Test
-	public void testReleaseExclusiveBuffers() throws Exception {
+	public void testConcurrentOnSenderBacklogAndRecycle() throws Exception {
 		// Setup
-		final SingleInputGate inputGate = mock(SingleInputGate.class);
-		final RemoteInputChannel inputChannel = createRemoteInputChannel(inputGate);
+		final NetworkBufferPool networkBufferPool = new NetworkBufferPool(256, 32, MemoryType.HEAP);
+		final ExecutorService executor = Executors.newFixedThreadPool(2);
+		final SingleInputGate inputGate = createSingleInputGate();
+		final RemoteInputChannel inputChannel  = createRemoteInputChannel(inputGate);
+		inputGate.setInputChannel(inputChannel.partitionId.getPartitionId(), inputChannel);
+		try {
+			final int numFloatingBuffers = 128;
+			final int numExclusiveSegments = 2;
+			final BufferPool bufferPool = networkBufferPool.createBufferPool(numFloatingBuffers, numFloatingBuffers);
+			inputGate.setBufferPool(bufferPool);
+			inputGate.assignExclusiveSegments(networkBufferPool, numExclusiveSegments);
+
+			// Exhaust all the floating buffers
+			final List<Buffer> floatingBuffers = new ArrayList<>(numFloatingBuffers);
+			for (int i = 0; i < numFloatingBuffers; i++) {
+				Buffer buffer = bufferPool.requestBuffer();
+				assertNotNull(buffer);
+				floatingBuffers.add(buffer);
+			}
 
-		// Assign exclusive segments to channel
-		final List<MemorySegment> exclusiveSegments = new ArrayList<>();
-		final int numExclusiveBuffers = 2;
-		for (int i = 0; i < numExclusiveBuffers; i++) {
-			exclusiveSegments.add(MemorySegmentFactory.allocateUnpooledSegment(1024, inputChannel));
+			// Exhaust all the exclusive buffers
+			final List<Buffer> exclusiveBuffers = new ArrayList<>(numExclusiveSegments);
+			for (int i = 0; i < numExclusiveSegments; i++) {
+				Buffer buffer = inputChannel.requestBuffer();
+				assertNotNull(buffer);
+				exclusiveBuffers.add(buffer);
+			}
+
+			final int backlog = 128;
+			final Callable<Void> requestBufferTask = new Callable<Void>() {
+				@Override
+				public Void call() throws Exception {
+					for (int j = 1; j <= backlog; j++) {
+						inputChannel.onSenderBacklog(j);
+					}
+
+					return null;
+				}
+			};
+
+			final Callable<Void> recycleBufferTask = new Callable<Void>() {
+				@Override
+				public Void call() throws Exception {
+					// Recycle all the exclusive buffers
```

Comment:
why not have two recycling threads, one for the exclusive buffers and another for the floating buffers? (you should then maybe increase the number of exclusive buffers)

Reply:
agree with it